storage/sql.go

```diff
@@ -499,3 +430,76 @@ func useNullIfEmpty(s string) *string {
 
 	return &s
 }
+
+func (d *Database) UnprocessedIndexedBlocks(ctx context.Context, maxHeight, limit int) (blocks.BlocksSynced, error) {
+	panic("REMOVED")
+}
+
+func (d *Database) MostRecentSyncedBlock(ctx context.Context) (*blocks.BlockSynced, error) {
+	panic("REMOVED")
+}
+
+func (d *Database) CollectAndMarkBlocksAsProcessing(ctx context.Context, batch int) (blocks.BlocksSynced, error) {
+	panic("REMOVED")
+}
+
+// LeaseGasOutputsMessages leases a set of messages that have receipts for gas output processing. minHeight and maxHeight define an inclusive range of heights to process.
+func (d *Database) LeaseGasOutputsMessages(ctx context.Context, claimUntil time.Time, batchSize int, minHeight, maxHeight int64) (derived.GasOutputsList, error) {
```

Comment:
This function contains a query that joins messages to receipts to only return messages that have corresponding receipts in the database. It also joins to block_headers to get the parent_base_fee. Is this the correct value to use?

Reply:
> This function contains a query that joins messages to receipts to only return messages that have corresponding receipts in the database

It should never be the case that a message without a receipt exists in the tables produced by visor since it operates on valid blocks, I believe this join on block_headers is unnecessary if its only for ensuring messages are returned that have a receipt. ([spec-ref](https://spec.filecoin.io/#systems__filecoin_vm__message__message-semantic-validation)) 

> It also joins to block_headers to get the parent_base_fee. Is this the correct value to use?

"The ParentBaseFee is used when calculating how much a sender burns when executing a message. Burning simply refers to sending attoFIL to a dedicated, unreachable account. A message causes ParentBaseFee * GasUsed attoFIL to be burnt" - ([lotus-ref](https://github.com/filecoin-project/lotus/blob/master/CHANGELOG.md#network-fee)). 

The `ParentBaseFee` value being extracted looks right to me :+1:

Reply:
I think the join between messages and receipts is necessary so as to ensure we only process messages that have received a receipt, which happens at some time in the future. The indexer will add messages and then some time later will find the receipts for those messages so there will be a period of time where we have messages with no receipts in the database.

Plus we need the gas_used value that is only in the receipt. The join to block headers it to retrieve the parent_base_fee. Both of these are needed for the call to ComputeGasOutputs.

Reply:
@frrist Thanks for the response. The join on `block_headers` is for returning `parent_base_fee`.

Reply:
BTW, this query looks good to me.